sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -124,11 +153,24 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
     val hiveVersion = externalCatalog.unwrapped.asInstanceOf[HiveExternalCatalog].client.version
     val stagingDir = hadoopConf.get("hive.exec.stagingdir", ".hive-staging")
     val scratchDir = hadoopConf.get("hive.exec.scratchdir", "/tmp/hive")
+    logDebug(s"path '${path.toString}', staging dir '$stagingDir', " +
+      s"scratch dir '$scratchDir' are used")
 
     if (hiveVersionsUsingOldExternalTempPath.contains(hiveVersion)) {
       oldVersionExternalTempPath(path, hadoopConf, scratchDir)
     } else if (hiveVersionsUsingNewExternalTempPath.contains(hiveVersion)) {
-      newVersionExternalTempPath(path, hadoopConf, stagingDir)
+      // HIVE-14270: Write temporary data to HDFS when doing inserts on tables located on S3
+      // Copied from Context.java#getTempDirForPath of Hive 2.3.
+      if (supportSchemeToUseNonBlobStore(path)) {
+        // Hive sets session_path as HDFS_SESSION_PATH_KEY(_hive.hdfs.session.path) in hive config
+        val HDFS_SESSION_PATH_KEY = "_hive.hdfs.session.path"
+        val sessionScratchDir = externalCatalog.unwrapped.asInstanceOf[HiveExternalCatalog]
+          .client.getConf(HDFS_SESSION_PATH_KEY, "")
+        logDebug(s"session scratch dir '$sessionScratchDir' is used")
+        getMRTmpPath(hadoopConf, sessionScratchDir, scratchDir)
```

Comment:
For the case of `hiveVersionsUsingNewExternalTempPath`, we use `stagingDir` for temp path. But if `supportSchemeToUseNonBlobStore` is true, we might use `scratchDir`.

I'm not sure about Hive's behavior. Is `scratchDir` still used after Hive 1.1?

Reply:
Yes, Hive's `scratchDir` is still used in Hive 2.0 or later.